### PR TITLE
chore(ui): basic human feedback rendering in calls table

### PR DESCRIPTION
**Pull request by gtarpenning**
## Description

MVP pr to combine feedback into the main calls data object. 

https://wandb.atlassian.net/browse/WB-21666

Screenshot 2024-11-18 at 10 05 26 AM

## Testing

How was this PR tested?

### CODE REVIEW

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallPage/cost/costUtils.ts [hunk 1]
```diff
@@ -87,22 +88,45 @@ export const formatTokenCost = (cost: number): string => {
...
+            ...call.traceCall?.summary,
+            weave: {
+              ...call.traceCall?.summary?.weave,
+              costs: costDict[call.callId].summary.weave.costs,
+            },
+          },
+          usage: costDict[call.callId].summary.usage,
+        },
+      };
+      return merged;
```
Comment (gtarpenning) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallPage/cost/costUtils.ts:129]: this looks more right... @jwlee64

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallPage/cost/costUtils.ts [hunk 2]
```diff
@@ -87,22 +88,45 @@ export const formatTokenCost = (cost: number): string => {
...
 export const addCostsToCallResults = (
   callResults: CallSchema[],
   costResults: CallSchema[]
-) => {
-  const costDict = costResults.reduce((acc, cost) => {
-    if (cost.callId) {
-      acc[cost.callId] = cost;
+): CallSchema[] => {
+  const costDict = costResults.reduce((acc, costResult) => {
+    if (costResult.callId) {
```
Comment (jwlee64) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallPage/cost/costUtils.ts:96]: if you wanted you could make this accumulator simpler 

```
acc[call.id] = {
  costs: costResult.traceCall?.summary?.weave?.costs,
  usage: costResult.traceCall?.summary?.usage,
  }
```
then make the code below

```
const merged = {
        ...call,
        traceCall: {
          ...call.traceCall,
          summary: {
            ...call.traceCall?.summary,
            weave: {
              ...call.traceCall?.summary?.weave,
              costs: costDict[call.callId].costs,
            },
          },
          usage: costDict[call.callId].usage,
        },
      };
```

but this looks right

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallPage/cost/costUtils.ts (hunk 1 above)
Comment (jwlee64) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallPage/cost/costUtils.ts:129]: nice, cost stuff lgtm, nit above

File: weave-js/src/components/PagePanelComponents/Home/Browse3.tsx [hunk 3]
```diff
@@ -736,7 +735,7 @@ const CallsPageBinding = () => {
     try {
       return JSON.parse(query.cols);
     } catch (e) {
-      return DEFAULT_COLUMN_VISIBILITY_CALLS;
+      return {};
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3.tsx:738]: why this change? Seems unrelated?

File: weave-js/src/components/PagePanelComponents/Home/Browse3/filters/common.ts [hunk 4]
```diff
@@ -40,6 +40,11 @@ export const FIELD_LABELS: Record<string, string> = {
 };
 
 export const getFieldLabel = (field: string): string => {
+  if (field.startsWith('feedback.')) {
+    // feedback.[wandb.annotation.foo].payload.value.foo --> feedback.foo
+    const bracketed = field.split('[')[1].split(']')[0];
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/filters/common.ts:45]: this parsing feels like something that should be centralized (alongside whatever constructs this)

File: weave-js/src/components/PagePanelComponents/Home/Browse3.tsx (hunk 3 above)
Comment (gtarpenning) [reply weave-js/src/components/PagePanelComponents/Home/Browse3.tsx:738]: This is actually related, we now have more fields we want to keep hidden, and they are dynamically generated so we can't rely on a constant. This handling is now moved into `callsTable`

File: weave-js/src/components/PagePanelComponents/Home/Browse3/filters/common.ts (hunk 4 above)
Comment (gtarpenning) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/filters/common.ts:45]: yeah good call, ill make a `parseFeedback` util and return a typed body

File: weave-js/src/components/PagePanelComponents/Home/Browse3.tsx (hunk 3 above)
Comment (tssweeney) [reply weave-js/src/components/PagePanelComponents/Home/Browse3.tsx:738]: i see - this was changed in the use effect

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableColumns.tsx [hunk 5]
```diff
@@ -327,6 +331,56 @@ function buildCallsTableColumns(
...
+      headerName: 'feedback',
+      children: feedbackCols.map(col => ({
+        field: convertFeedbackFieldToBackendFilter(col),
+      })),
+    });
+
+    const getFeedbackHeaderFromField = (field: string) => {
+      // feedback.wandb.annotation.Text-field-2.value.Text-field-2.ZQUEOy2FtgkRWahm8ucKgGHQGQYhoroXSug6SY6SSZQ
+      // -> Text-field-2
+      const regex = /feedback\.wandb\.[^.]+\.([^.]+)\.value\./;
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableColumns.tsx:360]: here is another thing that should be extracted

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableColumns.tsx [hunk 6]
```diff
@@ -327,6 +331,56 @@ function buildCallsTableColumns(
...
+    c.startsWith('feedback.')
+  );
+  if (feedbackCols.length > 0) {
+    const convertFeedbackFieldToBackendFilter = (field: string) => {
+      // feedback.wandb.annotation.Text-field-2.value.Text-field-2.ZQUEOy2FtgkRWahm8ucKgGHQGQYhoroXSug6SY6SSZQ
+      // -> feedback.[wandb.annotation.Text-field-2].value.Text-field-2.ZQUEOy2FtgkRWahm8ucKgGHQGQYhoroXSug6SY6SSZQ
+      const regex = /feedback\.wandb\.([^.]+\.[^.]+)\.value\./;
+      const feedbackType = regex.exec(field)?.[1] ?? field; // annotation.my-type
+      const afterType = field.split(feedbackType).pop();
+      return `feedback.[wandb.${feedbackType}]${afterType}`;
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableColumns.tsx:345]: same here

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableColumns.tsx [hunk 7]
```diff
@@ -327,6 +331,56 @@ function buildCallsTableColumns(
...
+    groupingModel.push({
+      groupId: 'feedback',
+      headerName: 'feedback',
+      children: feedbackCols.map(col => ({
+        field: convertFeedbackFieldToBackendFilter(col),
+      })),
+    });
+
+    const getFeedbackHeaderFromField = (field: string) => {
+      // feedback.wandb.annotation.Text-field-2.value.Text-field-2.ZQUEOy2FtgkRWahm8ucKgGHQGQYhoroXSug6SY6SSZQ
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableColumns.tsx:358]: isn't the format bracketed?

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableColumns.tsx [hunk 8]
```diff
@@ -327,6 +331,56 @@ function buildCallsTableColumns(
...
+      const regex = /feedback\.wandb\.[^.]+\.([^.]+)\.value\./;
+      const match = regex.exec(field);
+      return match?.[1] ?? field;
+    };
+
+    // Add feedback columns
+    const feedbackColumns: Array<GridColDef<TraceCallSchema>> =
+      feedbackCols.map(c => ({
+        field: convertFeedbackFieldToBackendFilter(c),
+        headerName: `Feedback.${getFeedbackHeaderFromField(c)}`,
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableColumns.tsx:369]: Making this comment here. Consider it a nit: It might be nice to have the group header for annotation say "Annotations"?

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableQuery.ts [hunk 9]
```diff
@@ -120,20 +125,87 @@ export const useCallsForQuery = (
...
+          },
+          {
+            $in: [
+              {$getField: 'weave_ref'},
+              callRefs.map(ref => ({$literal: ref})),
+            ],
+          },
+        ],
+      },
+      // TODO(gst): why does the $contains operator typing fail here...
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableQuery.ts:151]: still a todo?

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableQuery.ts [hunk 10]
```diff
@@ -120,20 +125,87 @@ export const useCallsForQuery = (
     callsStats.refetch();
   }, [calls, callsStats, costs]);
 
+  // query for annotation feeback
+  const {useFeedbackQuery} = useWFHooks();
+  const feedbackQueryQuery = useMemo(() => {
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableQuery.ts:130]: why not just do `include_feedback` and filter the results in the client?

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableQuery.ts [hunk 11]
```diff
@@ -120,20 +125,87 @@ export const useCallsForQuery = (
...
+    const callRefs = callResults.map(call =>
+      makeRefCall(entity, project, call.callId)
+    );
+    return {
+      $expr: {
+        $and: [
+          {
+            $contains: {
+              input: {$getField: 'feedback_type'},
+              substr: {$literal: `${ANNOTATION_FEEDBACK_TYPE_PREFIX}.`},
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableQuery.ts:140]: annotation type currently, but what about other feedbacks?

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableQuery.ts (hunk 10 above)
Comment (gtarpenning) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableQuery.ts:130]: I tried both, I thought this way would be more performant: we can get a faster first-paint, then fill in feedback

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableQuery.ts (hunk 11 above)
Comment (gtarpenning) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableQuery.ts:140]: we can add to this query, or make the prefix less restrictive

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableQuery.ts (hunk 10 above)
Comment (tssweeney) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableQuery.ts:130]: i see. makes sense. I think the feedback type filter might be better to do client side since:
1. The data size for feedback is quite small per row (so bandwidth is not a concern as much)
2. You will want to do more than just annotation really soon

Comment (gtarpenning) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableQuery.ts:130]: yep that makes sense ez pivot

File: weave-js/src/components/PagePanelComponents/Home/Browse3/feedback/HumanFeedback/tsHumanFeedback.ts [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+  // userDefinedType: Numerical-field-2
+  // type: annotation
+  // displayName: Annotation.Numerical-field-2
+
+  // If the field is coming from the flattened table, remove the
+  // summary portion
+  const field = inputField.startsWith('summary.weave.feedback')
+    ? inputField.replace('summary.weave.feedback.', '')
+    : inputField;
+  const deBracketed = field.replace(/\[.*\]/g, '');
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/feedback/HumanFeedback/tsHumanFeedback.ts:25]: the fact that this is looking for a bracket does not make sense to me given the comment above

Comment (tssweeney) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/feedback/HumanFeedback/tsHumanFeedback.ts:25]: i think this is just a no-op right now

**Review by tssweeney — APPROVED** [review]

File: weave-js/src/components/PagePanelComponents/Home/Browse3/feedback/HumanFeedback/tsHumanFeedback.ts (hunk 12 above)
Comment (gtarpenning) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/feedback/HumanFeedback/tsHumanFeedback.ts:25]: ill update the comment, but I am using this fn in one place where we go from backend filter -> displayName 

```javascript
export const getFieldLabel = (field: string): string => {
  if (field.startsWith('feedback.')) {
    // Here the field is coming from convertFeedbackFieldToBackendFilter
    // so the field should start with 'feedback.' if feedback
    const parsed = parseFeedbackType(field);
    if (parsed === null) {
      return field;
    }
    return parsed.displayName;
  }
  return FIELD_LABELS[field] ?? field;
};
```

Comment (tssweeney) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/feedback/HumanFeedback/tsHumanFeedback.ts:25]: hmmm, i would probably split those, but sure. It seems like a sufficiently different case

Comment (tssweeney) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/feedback/HumanFeedback/tsHumanFeedback.ts:25]: hmmm, i would probably split those, but sure. It seems like a sufficiently different case

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableQuery.ts [hunk 13]
```diff
@@ -108,11 +109,12 @@ export const useCallsForQuery = (
     {
       skip: calls.loading,
       includeCosts: true,
+      includeFeedback: true,
```
Comment (gtarpenning) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CallsPage/callsTableQuery.ts:112]: `includeFeedback: true,`
